[git-webkit] Cache filed bug

JonWBedard opened the pull request:
#### 2425de1c879b13f96a49d4b2cb83d587d73e536b
<pre>
[git-webkit] Cache filed bug
<a href="https://bugs.webkit.org/show_bug.cgi?id=255007">https://bugs.webkit.org/show_bug.cgi?id=255007</a>
rdar://107629444

Reviewed by Elliott Williams.

git-webkit branch will file a bug. However, that bug is only picked up
in a commit message through environment variables, which means that
bugs filed by `git-webkit branch` aren&apos;t very useful except when happening
through `git-webkit pr`&apos;s invocation of `git-webkit branch`. We have a
similar problem when crafting cherry-picks or reverting changes.

* Tools/Scripts/hooks/prepare-commit-msg: Pull title, bug and cherry-pick message from
`git config` if they aren&apos;t available in the environment.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py: Add ability to get
and set `git config` values.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.main): Save filed bug into `git config` environment.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/cherry_pick.py:
(CherryPick.main): Save cherry-pick pattern and bug details into `git config`, use
COMMIT_MESSAGE_BUG instead of GIT_WEBKIT_BUG.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/command.py:
(Command.write_branch_variables): Write variable to `git config`&apos;s branch specific namespace.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.revert_commit): Save filed bug into `git config` environment.

Canonical link: <a href="https://commits.webkit.org/263641@main">https://commits.webkit.org/263641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb5201fb4aa515f0b01ca342183802289951e753

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5326 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5651 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6863 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/5368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5444 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5425 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/5498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/4771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6889 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/5431 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/2978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/4776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/4844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/6470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5281 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/5275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/4747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/8843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/597 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/5107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -11,26 +12,39 @@ SCRIPTS = os.path.dirname(os.path.dirname(LOCATION))
 from webkitbugspy import radar
 
 
+cwd = ENVIRONMENT.get('PWD', os.getcwd())
```
What motivates allowing `PWD` to override what the system believes the cwd to be?

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -11,26 +12,39 @@ SCRIPTS = os.path.dirname(os.path.dirname(LOCATION))
 
 
+cwd = ENVIRONMENT.get('PWD', os.getcwd())
+environment_file = os.path.join(cwd, '.git', 'commit-message-environment.json')
```
Given this is a hook, shouldn't we use `ENVIRONMENT['GIT_DIR']` here?

JonWBedard replied:
`GIT_DIR` is not passed by `git`, playing around a bit more, I expected this information to be passed by `PWD`, but `PWD` is the working directory of the calling command, `os.getcwd()` will always be the repository, it seems.

gsnedders replied:
https://git-scm.com/docs/githooks certainly claims that in a hook:

> Environment variables, such as GIT_DIR, GIT_WORK_TREE, etc., are exported so that Git commands run by the hook can correctly locate the repository.

JonWBedard replied:
Maybe they are only exported if they are set? Don't see those when dumping the environment in our `prepare-commit-msg` hook.

JonWBedard replied:
We don't need any of this with `git rev-parse --absolute-git-dir`, `git` will figure out everything from there, it would seem.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/command.py:
```diff
@@ -36,6 +37,31 @@ class Command(object):
+    def commit_message_environment_path(cls, repository):
+        if not repository:
+            return None
+        return os.path.join(repository.root_path, '.git', 'commit-message-environment.json')
```
This won't work in a worktree where `.git` is a file; we should probably get `$GIT_DIR` here instead. (We have `repository.common_directory` which gives `$GIT_COMMON_DIR`, but we don't seem to have a `repository.git_dir` or similar.)

JonWBedard replied:
Seems like we should just use `common_directory`? I struggle to think of a circumstance where that would be wrong.

gsnedders replied:
With worktrees you can have multiple commits in progress at once, so you could have two partial commits (e.g., waiting on $EDITOR to exit), and therefore we need to have `commit-message-environment.json` in per-worktree directories.

JonWBedard replied:
Ah, so we do need to use `GIT_DIR` where as identifiers (which are the other primary consumer here) need to use `GIT_COMMON_DIR`

gsnedders reviewed (commented):
(non-reviewer review:)

While broadly good, this seems to add further breakage in worktrees. See below:

emw-apple commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -25,20 +26,43 @@ from webkitpy.common.checkout.diff_parser import DiffParser
+        env_from_config[mapped_key] = value
+for key, value in env_from_config.items():
+    if key not in ENVIRONMENT:
+        ENVIRONMENT[key] = value
```
This setup (loading arbitrary environment variables from git-config) seems quite general, is there future use you’re anticipating other than “pull two config settings when populating the commit message”?

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -308,6 +308,63 @@ def to_identifier(self, hash=None, revision=None, populate=True, branch=None):
+        def __init__(self, repository):
+            self._repository = repository
+
+        def cleanup(self):
```
With the changes we talked about in person (using the same `branch.*` config namespace and removing the HEAD behavior), we can delete this and rely on git’s built-in branch cleanup logic.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/cherry_pick.py:
```diff
@@ -89,7 +89,14 @@ def main(cls, args, repository, **kwargs):
+        env['COMMIT_MESSAGE_BUG'] = issue.link if issue else ''
+
+        repository.branch_environment.cleanup()
+        repository.branch_environment.set(dict(
```
Is it necessary to do the git-config dance when cherry picking? The cherry-pick should either succeed or fail with conflicts, and in the latter case the prepared commit message should be saved.

JonWBedard replied:
TLDR, yes.

The most common cherry-picking pattern is defining a different radar (probably a clone of the original) and tracking the cherry-pick with that radar. In that case, we'd be defining a different "COMMIT_MESSAGE_BUG" than the original commit message has. If we were interrupted by a conflict resolution and didn't cache that value somehow, we would lose it. This is particularly relevant for combining this change with #12058. We will likely have a `--clone` option on our `cherry-pick` command that will do the cherry-picking and cloning in a single script invocation.

emw-apple replied:
> If we were interrupted by a conflict resolution and didn't cache that value somehow, we would lose it.

OK, makes sense. I didn't realize that during conflict resolution, .git/COMMIT_EDITMSG is the _original_ commit message, and `prepare-commit-msg` is not called until conflicts are fixed and you run `git cherry-pick --continue`.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -114,6 +114,13 @@ def revert_commit(cls, args, repository, **kwargs):
         env['COMMIT_MESSAGE_BUG'] = '\n'.join(bug_urls)
+
+        repository.branch_environment.cleanup()
+        repository.branch_environment.set(dict(
```
Ditto.

emw-apple reviewed (commented):
Thanks again for talking irl about this! Here are a few small questions, but mostly just curious to see the next revision.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -308,6 +308,63 @@ def to_identifier(self, hash=None, revision=None, populate=True, branch=None):
+                cwd=self._repository.path,
+                encoding='utf-8',
+            ).stdout.splitlines():
+                full_key, value = line.split(maxsplit=1)
```
Do we really not have any generic code to do this? Because this (and in `prepare-commit-msg` above) does seem somewhat fragile—they rely on the values to not contain new lines (c.f. `--null`) and for the config file to be valid utf-8 (which, to be fair, it probably will be… but git doesn't actually guarantee/enforce that).

Not gonna say we need to fix this before we land _this_, but it would be good to add something before this gets worse.

JonWBedard replied:
Not ignoring this feedback, but the change I just uploaded essentially makes it obsolete.

We do have generic code to do this, but I was trying to limit the scope of `config` values we were extracting. Also, having a hook share code with the repository it's operating on is somewhat unreliable, so I try to avoid it where possible, hence the duplication in that change. New change uses a different approach, so there won't be duplication.

emw-apple commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -11,6 +11,7 @@ SCRIPTS = os.path.dirname(os.path.dirname(LOCATION))
 CHERRY_PICKING_RE = re.compile(r'^# You are currently cherry-picking commit (?P<hash>[a-f0-9A-F]+)\.$')
 IDENTIFIER_RE = re.compile(r'(\d+\.)?\d+@\S+')
 PREFER_RADAR = {{ prefer_radar }}
+ENVIRONMENT = os.environ.copy()
```
nit: Why do we need to avoid mutating the actual environment?

JonWBedard replied:
We don't. This was originally done because merging dictionaries is easier than `os.environ` (which is dictionary-like, but not exactly a dictionary), but that isn't true any more.

JonWBedard committed:
[git-webkit] Cache filed bug
https://bugs.webkit.org/show_bug.cgi?id=255007
rdar://107629444

Reviewed by Elliott Williams.

git-webkit branch will file a bug. However, that bug is only picked up
in a commit message through environment variables, which means that
bugs filed by `git-webkit branch` aren't very useful except when happening
through `git-webkit pr`'s invocation of `git-webkit branch`. We have a
similar problem when crafting cherry-picks or reverting changes.

* Tools/Scripts/hooks/prepare-commit-msg: Pull title, bug and cherry-pick message from
`git config` if they aren't available in the environment.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py: Add ability to get
and set `git config` values.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.main): Save filed bug into `git config` environment.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/cherry_pick.py:
(CherryPick.main): Save cherry-pick pattern and bug details into `git config`, use
COMMIT_MESSAGE_BUG instead of GIT_WEBKIT_BUG.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/command.py:
(Command.write_branch_variables): Write variable to `git config`'s branch specific namespace.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.revert_commit): Save filed bug into `git config` environment.

Canonical link: https://commits.webkit.org/263641@main
